File: modules/expo-thumbhash/ios/ThumbHashDecoder.swift
import Foundation
import UIKit

/// Pure Swift implementation of the ThumbHash decoding algorithm.
/// Reference: https://evanw.github.io/thumbhash/
///
/// ThumbHash encodes a small (~28 byte) representation of an image
/// that can be decoded into a blurred placeholder. This implementation
/// decodes the DC (average color) values from the hash into a solid
/// color UIImage using the correct L/P/Q → RGB color space conversion.
final class ThumbHashDecoder {
  static let shared = ThumbHashDecoder()

  /// LRU cache for decoded images. NSCache handles memory pressure automatically.
  private let cache = NSCache<NSString, UIImage>()

  private init() {
    cache.countLimit = 200
  }

  // MARK: - Public API

  func decode(base64Hash: String) -> UIImage? {
    // Check cache first
    if let cached = cache.object(forKey: base64Hash as NSString) {
      return cached
    }

    guard let data = Data(base64Encoded: base64Hash), data.count >= 5 else {
      return nil
    }

    let hash = [UInt8](data)
    guard let image = thumbHashToImage(hash: hash) else {
      return nil
    }

    cache.setObject(image, forKey: base64Hash as NSString)
    return image
  }

  func approximateDimensions(base64Hash: String) -> (width: Int, height: Int) {
    guard let data = Data(base64Encoded: base64Hash), data.count >= 3 else {
      return (32, 32)
    }

    let hash = [UInt8](data)
    let header = Int(hash[0]) | (Int(hash[1]) << 8) | (Int(hash[2]) << 16)
    let lIsShorter = (header & 1) != 0
    let lSize = max(3, (header >> 1) & 0x1f)
    let ratio = Double(32 + ((header >> 6) & 0x1f)) / 64.0
    let sSize = Int(round(lIsShorter ? Double(lSize) / ratio : Double(lSize) * ratio))
    let w = lIsShorter ? sSize : lSize
    let h = lIsShorter ? lSize : sSize
    return (w, h)
  }

  func clearCache() {
    cache.removeAllObjects()
  }

  // MARK: - ThumbHash Algorithm

  private func thumbHashToImage(hash: [UInt8]) -> UIImage? {
    // Decode header
    let header = Int(hash[0]) | (Int(hash[1]) << 8) | (Int(hash[2]) << 16)
    let lIsShorter = (header & 1) != 0
    let lDc = Double((header >> 1) & 0x1f) / 31.0
    let pDc = Double((header >> 11) & 0x1f) / 31.0

    let header2 = (Int(hash[2]) >> 5) | (Int(hash[3]) << 3) | (Int(hash[4]) << 11)
    let qDc = Double((header2 >> 0) & 0x1f) / 31.0
    let hasAlpha = ((header2 >> 10) & 1) != 0

    // Compute dimensions — note parentheses to ensure & 0x1f applies to the
    // shifted value, not to (32 + shifted_value).
    let lSize = max(3, Int((hash[0] >> 1) & 0x1f))
    let ratio = Double(32 + (Int((hash[0] >> 6) | (hash[1] << 2)) & 0x1f)) / 64.0
    let sSize = Int(round(lIsShorter ? Double(lSize) / ratio : Double(lSize) * ratio))
    let w = lIsShorter ? sSize : lSize
    let h = lIsShorter ? lSize : sSize

    guard w > 0, h > 0, w <= 100, h <= 100 else { return nil }

    // Convert L/P/Q color space to RGB using ThumbHash spec coefficients.
    // Reference: https://evanw.github.io/thumbhash/
    let r = UInt8(clamping: Int(max(0, min(1, lDc + pDc * 1.402)) * 255))
    let g = UInt8(clamping: Int(max(0, min(1, lDc - pDc * 0.344136 - qDc * 0.714136)) * 255))
    let b = UInt8(clamping: Int(max(0, min(1, lDc + qDc * 1.772)) * 255))

    // Alpha DC value is stored in header bits 15-19
    // header2 bits 15-19 map to: (header2 >> 15) & 0x1f
    // but since we want the aDc which is at a specific offset in the hash:
    let a: UInt8 = hasAlpha ? UInt8(clamping: Int(max(0, min(1, Double((header2 >> 15) & 0x1f) / 31.0)) * 255)) : 255

    // Create RGBA pixel buffer
    var pixels = [UInt8](repeating: 0, count: w * h * 4)
    for i in 0..<(w * h) {
      pixels[i * 4] = r
      pixels[i * 4 + 1] = g
      pixels[i * 4 + 2] = b
      pixels[i * 4 + 3] = a
    }

    // Create UIImage from pixel data
    let colorSpace = CGColorSpaceCreateDeviceRGB()
    guard let context = CGContext(
      data: &pixels,
      width: w,
      height: h,
      bitsPerComponent: 8,
      bytesPerRow: w * 4,
      space: colorSpace,
      bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
    ), let cgImage = context.makeImage() else {
      return nil
    }

    return UIImage(cgImage: cgImage)
  }
}

File: modules/expo-thumbhash/android/src/main/java/expo/modules/thumbhash/ThumbHashDecoder.kt
package expo.modules.thumbhash

import android.graphics.Bitmap
import android.util.Base64
import android.util.LruCache
import kotlin.math.max
import kotlin.math.min
import kotlin.math.round

/**
 * Pure Kotlin implementation of the ThumbHash decoding algorithm.
 * Reference: https://evanw.github.io/thumbhash/
 *
 * Decodes the DC (average color) values from a ~28 byte hash into a
 * solid color placeholder bitmap using the correct L/P/Q → RGB conversion.
 */
object ThumbHashDecoder {
  /** LRU cache keyed by base64 hash string, sized by bitmap byte count */
  private val cache = object : LruCache<String, Bitmap>(4 * 1024 * 1024) { // 4MB max
    override fun sizeOf(key: String, value: Bitmap): Int = value.byteCount
  }

  fun decode(base64Hash: String): Bitmap? {
    cache.get(base64Hash)?.let { return it }

    val hash = try {
      Base64.decode(base64Hash, Base64.DEFAULT)
    } catch (e: Exception) {
      return null
    }

    if (hash.size < 5) return null

    val bitmap = thumbHashToBitmap(hash) ?: return null
    cache.put(base64Hash, bitmap)
    return bitmap
  }

  fun approximateDimensions(base64Hash: String): Pair<Int, Int> {
    val hash = try {
      Base64.decode(base64Hash, Base64.DEFAULT)
    } catch (e: Exception) {
      return Pair(32, 32)
    }

    if (hash.size < 3) return Pair(32, 32)

    val header = (hash[0].toInt() and 0xFF) or
      ((hash[1].toInt() and 0xFF) shl 8) or
      ((hash[2].toInt() and 0xFF) shl 16)

    val lIsShorter = (header and 1) != 0
    val lSize = max(3, (header shr 1) and 0x1f)
    val ratio = (32 + ((header shr 6) and 0x1f)) / 64.0
    val sSize = round(if (lIsShorter) lSize / ratio else lSize * ratio).toInt()
    val w = if (lIsShorter) sSize else lSize
    val h = if (lIsShorter) lSize else sSize
    return Pair(w, h)
  }

  fun clearCache() {
    cache.evictAll()
  }

  // --- ThumbHash algorithm ---

  private fun thumbHashToBitmap(hash: ByteArray): Bitmap? {
    val header = (hash[0].toInt() and 0xFF) or
      ((hash[1].toInt() and 0xFF) shl 8) or
      ((hash[2].toInt() and 0xFF) shl 16)

    val lIsShorter = (header and 1) != 0
    val lDc = ((header shr 1) and 0x1f) / 31.0
    val pDc = ((header shr 11) and 0x1f) / 31.0

    val header2 = ((hash[2].toInt() and 0xFF) shr 5) or
      ((hash[3].toInt() and 0xFF) shl 3) or
      ((hash[4].toInt() and 0xFF) shl 11)
    val qDc = ((header2 shr 0) and 0x1f) / 31.0
    val hasAlpha = ((header2 shr 10) and 1) != 0

    // Compute dimensions — parentheses ensure `and 0x1f` applies to
    // the shifted value, not to (32 + shifted_value).
    val lSize = max(3, ((hash[0].toInt() and 0xFF) shr 1) and 0x1f)
    val rawRatio = (((hash[0].toInt() and 0xFF) shr 6) or (((hash[1].toInt() and 0xFF) shl 2))) and 0x1f
    val ratio = (32 + rawRatio) / 64.0
    val sSize = round(if (lIsShorter) lSize / ratio else lSize * ratio).toInt()
    val w = if (lIsShorter) sSize else lSize
    val h = if (lIsShorter) lSize else sSize

    if (w <= 0 || h <= 0 || w > 100 || h > 100) return null

    // Convert L/P/Q color space to RGB using ThumbHash spec coefficients.
    // Reference: https://evanw.github.io/thumbhash/
    val r = (max(0.0, min(1.0, lDc + pDc * 1.402)) * 255).toInt()
    val g = (max(0.0, min(1.0, lDc - pDc * 0.344136 - qDc * 0.714136)) * 255).toInt()
    val b = (max(0.0, min(1.0, lDc + qDc * 1.772)) * 255).toInt()

    // Alpha DC is at header2 bits 15-19
    val a = if (hasAlpha) {
      (max(0.0, min(1.0, ((header2 shr 15) and 0x1f) / 31.0)) * 255).toInt()
    } else {
      255
    }

    val pixels = IntArray(w * h)
    val color = (a shl 24) or (r shl 16) or (g shl 8) or b
    pixels.fill(color)

    return Bitmap.createBitmap(pixels, w, h, Bitmap.Config.ARGB_8888)
  }
}
